tasmota/xsns_96_qmc5883.ino

```diff
@@ -0,0 +1,217 @@
+  Wire.endTransmission();
+  QMC5883.MX=(int)(int16_t)(buff[0] | buff[1] << 8);
+  QMC5883.MZ=(int)(int16_t)(buff[2] | buff[3] << 8);
+  QMC5883.MY=(int)(int16_t)(buff[4] | buff[5] << 8);
+  */
+  QMC5883.MX = (int)(int16_t)(Wire.read() | Wire.read() << 8);
+  QMC5883.MY = (int)(int16_t)(Wire.read() | Wire.read() << 8);
+  QMC5883.MZ = (int)(int16_t)(Wire.read() | Wire.read() << 8);
+  byte overflow = Wire.read() & 0x02;
+  
+  overflow = overflow << 2;
+  
+
+  // Hold the module so that Z is pointing 'up' and you can measure the heading with x&y
+  // Calculate heading when the magnetometer is level, then correct for signs of axis.
+  float heading=atan2((int)QMC5883.MY, (int)QMC5883.MX) * 180.0/M_PI;
+
+  if (heading < 0)
+    heading += 360.0;
+
+  /*
+  // Check for wrap due to addition of declination.
+  if(heading > 2*PI)
+    heading -= 2*PI;
+  */
+  // Convert radians to degrees for readability.
+  QMC5883.HG = (int)heading; 
+
+}
+
+
+/*********************************************************************************************\
+ * Presentation
+\*********************************************************************************************/
+
+#ifdef USE_WEBSERVER
+const char HTTP_SNS_QMC5883[] PROGMEM =
+  "{s}QMC5883 " D_MX "{m}%d " D_UNIT_MICROTESLA  "{e}"      // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+  "{s}QMC5883 " D_MY "{m}%d " D_UNIT_MICROTESLA  "{e}"      // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+  "{s}QMC5883 " D_MZ "{m}%d " D_UNIT_MICROTESLA  "{e}"      // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+  "{s}QMC5883 " D_HG "{m}%d " D_UNIT_DEGREE      "{e}";     // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+
+const char HTTP_SNS_QMC5883_ERROR[] PROGMEM =
+  "{s}QMC5883 {m} %s {e}";
+#endif
+
+
+void QMC5883_Show(uint8_t json)
+{
+  QMC5883_read_data();
+
+  if (json) {
+    if (!QMC5883.ready){
+      AddLog(LOG_LEVEL_INFO, PSTR("QMC5883: " D_ERROR " %x" ),QMC5883.status);
+      return;
+  }
+  else {
+    ResponseAppend_P(PSTR(",\"QMC5883\":{\"" D_JSON_MX "\":%i,\"" D_JSON_MY "\":%i,\"" D_JSON_MZ "\":%i,\"" D_JSON_HEADING "\":%i}"), QMC5883.MX, QMC5883.MY, QMC5883.MZ, QMC5883.HG);
+    }
+#ifdef USE_WEBSERVER
+    } else {
+      switch(QMC5883.ready){
+        case true:
+          WSContentSend_PD(HTTP_SNS_QMC5883, QMC5883.MX, QMC5883.MY, QMC5883.MZ, QMC5883.HG);
+          break;
+        case false:
+          WSContentSend_PD(HTTP_SNS_QMC5883_ERROR, D_START);
+          break;
+        default:
+          WSContentSend_PD(HTTP_SNS_QMC5883_ERROR, D_ERROR);
+      }
+#endif
+  }
+}
+
+
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+ 
+bool Xsns96(byte function)
+{
+  if (!I2cEnabled(XI2C_66)) { return false; }
+
+  bool result = false;
+
+  if (FUNC_INIT == function) {
+    QMC5883_Init();
+  }
+  else if (QMC5883.ready) {

```

Comment:
Similar, look at the HMC5883 driver. Use the dynamic pointer for that purpose instead of wasting a byte